[TG-9299] Fix crash in case of user provided clinit for a clinit which is not in the symbol table

romainbrenguier opened the pull request:
The look-up for clinit could fail when we had a user provided clinit but the real clinit was not in the symbol table.
This adds a test which illustrate the case that was previously causing the exception.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Replace lookup_ref call by lookup

This code doesn't need a writable reference so it should call lookup
instead.
We also add an invariant to check that the reference does exists, this
gives more helpful messages than the generic `_Map_base::at`.

romainbrenguier committed:
Do not call assign_from_json if not in symbol_table

The assign_from_json will fail with an invariant failure if it is called
on a method name that does not exist in the symbol table so we should be
careful when calling it.
This also make the function return early when the real clinit is not in
the symbol table, which can avoid some unecessary computation.

peterschrammel commented on jbmc/regression/jbmc/deterministic_assignments_json/test_empty_non_empty.desc:
```diff
@@ -0,0 +1,12 @@
+--
+This tests what happens when the non-empty class is substitued with an empty
+one after compilation: After compiling EmptyNonEmpty.class and Empty.class,
+Empty.class is moved to NonEmpty.class.
```
What does 'moved' mean?

romainbrenguier replied:
I've clarified that

peterschrammel commented on jbmc/regression/jbmc/deterministic_assignments_json/EmptyNonEmpty.java:
```diff
@@ -0,0 +1,9 @@
+public class EmptyNonEmpty {
+
+  /// This test is meant to illustrate what happens when the non-empty
+  /// class is substitued with an empty one after compilation.
```
substituted

peterschrammel commented on jbmc/regression/jbmc/deterministic_assignments_json/test_empty_non_empty.desc:
```diff
@@ -0,0 +1,12 @@
+^SIGNAL=0$
+--
+--
+This tests what happens when the non-empty class is substitued with an empty
```
substituted

smowton commented on jbmc/regression/jbmc/deterministic_assignments_json/Empty.java:
```diff
@@ -0,0 +1,3 @@
+public class Empty {
```
Rather that emptiness, is having or not having static fields the critical attribute?

romainbrenguier replied:
Yes, I've clarified that in the class names

smowton commented on jbmc/src/java_bytecode/assignments_from_json.cpp:
```diff
@@ -838,7 +838,9 @@ void assign_from_json(
   allocate_objectst allocate(ID_java, location, function_id, symbol_table);
   code_blockt body_rec;
-  const auto class_name = declaring_class(symbol_table.lookup_ref(function_id));
+  const symbolt *function_symbol = symbol_table.lookup(function_id);
```
is lookup-ref vs. lookup-then-invariant really that different?

peterschrammel replied:
yes, you get a reasonable message on failure.

smowton commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -808,13 +809,15 @@ code_blockt get_user_specified_clinit_body(
           }
         }
+        if(symbol_table.lookup(real_clinit_name) == nullptr)
+          return code_blockt{};
```
Add a comment here about how/why this could happen

romainbrenguier replied:
done

peterschrammel commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -777,6 +777,7 @@ code_blockt get_user_specified_clinit_body(
   std::unordered_map<std::string, object_creation_referencet> &references)
 {
   jsont json;
+  const irep_idt real_clinit_name = clinit_function_name(class_id);
```
(commit message) careful

smowton commented on jbmc/regression/jbmc/deterministic_assignments_json/EmptyNonEmpty.java:
```diff
@@ -0,0 +1,9 @@
+public class EmptyNonEmpty {
```
Change this name too

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81e1c9b1b6ca20f7d32da67410d135f3d72530cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/125753721

romainbrenguier committed:
Drop intermediary reference to keep only one if

The declaration of the reference class_json_value prevented merging the
two if, and it can be removed without making the code more complicated
since it is just an alias for class_entry->second

romainbrenguier committed:
Unit test for user specified clinit when there is no clinit

Test what happens when clinit for a class is not in the symbol table
though the JSON file contains entries for this class.
This used to cause map::at to throw an exception.

antlechner commented on jbmc/src/java_bytecode/assignments_from_json.cpp:
```diff
@@ -838,7 +838,9 @@ void assign_from_json(
   code_blockt body_rec;
-  const auto class_name = declaring_class(symbol_table.lookup_ref(function_id));
+  const symbolt *function_symbol = symbol_table.lookup(function_id);
+  INVARIANT(function_symbol, "Function must appear in symbol table");
```
:bulb: This message is quite general, maybe include `function_id` in it and/or mention the context in which this was looked up.

romainbrenguier replied:
I don't know, the backtrace of the invariant should already give us the context.

antlechner commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -817,6 +817,13 @@ code_blockt get_user_specified_clinit_body(
+      if(symbol_table.lookup(real_clinit_name) == nullptr)
+      {
+        // Case where there is a user provided clinit, but the real clinit
+        // doesn't appear in the symbol table. This may occur when some class
```
:question: Do you have an example of such a class substitution? Having these additional checks in is good of course, but I don't understand how this ever happened in practice. `get_user_specified_clinit_body` is only called to populate the body of a `synthetic_method_typet::USER_SPECIFIED_STATIC_INITIALIZER` if we see one. Such a synthetic method is only created in `create_user_specified_clinit_function_symbol`, which in turn is only called after several `if`s, one of which is `symbol_table.has_symbol(clinit_function_name(class_name))`.
So it looks like the only way we could end up with a `user_specified_clinit` but no `clinit` would be if `clinit` was first added to the symbol table, then `user_specified_clinit` was added, and finally `clinit` was removed - but I'm not aware of the latter ever happening?

romainbrenguier replied:
@antlechner 
`create_user_specified_clinit_function_symbol` is called if `needs_clinit_wrapper` is true, which is when `clinit` exists *or* some other condition. If you think this is wrong then `needs_clinit_wrapper` needs to be corrected.

antlechner replied:
Ah, that's the part I overlooked! `needs_clinit_wrapper` is true if `clinit` exists for this class _or_ a superclass.
So it looks like the mistake is that currently, we create a `user_specified_clinit` iff we create a `clinit_wrapper`. But really we should be creating a `user_specified_clinit` iff there is a `clinit`.

romainbrenguier replied:
ok, I'm still going to merge this PR since it prevents a crash but the condition for calling `create_user_specified_clinit_function_symbol` should be fixed. Do you want to create an issue for that?

antlechner commented on jbmc/src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -817,6 +817,13 @@ code_blockt get_user_specified_clinit_body(
           }
         }
       }
+      if(symbol_table.lookup(real_clinit_name) == nullptr)
```
:bulb: Looks like it would make more sense to move this check just below the definition of `real_clinit_name`.
Edit: Ah, this is done in the next commit. Those two commits should probably be squashed.

antlechner reviewed (approved):
Is there any way to add a regression test for this? See my question below: I can't see how the bug that this is fixing would occur in practice.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1b3c289e208651bdf5bcb74b8dd29299149e8b0e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/126241524
